### PR TITLE
[set-keychain-accessibility] - accessibility;

### PR DESCRIPTION
**Pull request by cladvd**
Integrated the definition of accessibility level for the keychain of iOS

### CODE REVIEW

**Comment by gaebel** [comment]
Hi @cladvd, thanks for the contribution :) I have a few remarks regarding the implementation:

File: src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 1]
```diff
@@ -252,15 +260,18 @@ actual open class KVault(
     private fun <T> context(vararg values: Any?, block: Context.(List<CFTypeRef?>) -> T): T {
         val standard = mapOf(
             kSecAttrService to CFBridgingRetain(serviceName),
-            kSecAttrAccessGroup to CFBridgingRetain(accessGroup)
+            kSecAttrAccessGroup to CFBridgingRetain(accessGroup),
+            kSecAttrAccessible to CFBridgingRetain(accessibility)
```
Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KVault.kt:264]:
```suggestion
            kSecAttrAccessible to CFBridgingRetain(accessibility?.value)
```

File: src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 2]
```diff
@@ -13,11 +13,13 @@ import platform.darwin.noErr
...
  * @param serviceName Name of the service. Used to categories entries.
  * @param accessGroup Name of the access group. Used to share entries between apps.
+ * @param accessibility Level of the accessibility for the Keychain instance.
  * @constructor Initiates a Keychain with the given parameters.
  */
 actual open class KVault(
     val serviceName: String? = null,
-    val accessGroup: String? = null
+    val accessGroup: String? = null,
+    val accessibility: KSecAttrAccessible? = KSecAttrAccessible.WhenUnlocked
```
Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KVault.kt:22]: Not needed to define `KSecAttrAccessible` as optional.

File: src/iosMain/kotlin/com/liftric/kvault/KSecAttrAccessible.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import platform.Security.*
+
+/**
+ * kSecAttrAccessible attributes wrapper.
+ * attribute enables you to control item availability relative to the lock state of the device.
+ * It also lets you specify eligibility for restoration to a new device.
+ * If the attribute ends with the string ThisDeviceOnly, the item can be restored to the same device
+ * that created a backup, but it isn’t migrated when restoring another device’s backup data.
+ */
+enum class KSecAttrAccessible(val value: CFStringRef?) {
```
Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KSecAttrAccessible.kt:13]: Maybe rename to `Accessible` and put it in the `KVault` class for scoping => `KVault.Accessible`

File: src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 4]
```diff
@@ -142,8 +159,9 @@ actual open class KVault(
         val query = query(
             kSecClass to kSecClassGenericPassword,
             kSecAttrAccount to account,
-            kSecReturnData to kCFBooleanFalse
-        )
+            kSecReturnData to kCFBooleanFalse,
+
```
Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KVault.kt:163]:
```suggestion
```

File: src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 5]
```diff
@@ -222,7 +242,8 @@ actual open class KVault(
             kSecClass to kSecClassGenericPassword,
             kSecAttrAccount to account,
             kSecReturnData to kCFBooleanTrue,
-            kSecMatchLimit to kSecMatchLimitOne
+            kSecMatchLimit to kSecMatchLimitOne,
+            kSecAttrAccessible to accessibility.value
```
Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KVault.kt:246]:
```suggestion
```

File: src/iosMain/kotlin/com/liftric/kvault/KVault.kt [hunk 6]
```diff
@@ -195,18 +213,20 @@ actual open class KVault(
...
+
     }
 
     private fun update(key: String, value: Any?): Boolean = context(key, value) { (account, data) ->
         val query = query(
             kSecClass to kSecClassGenericPassword,
             kSecAttrAccount to account,
-            kSecReturnData to kCFBooleanFalse
+            kSecReturnData to kCFBooleanFalse,
+            kSecAttrAccessible to  accessibility.value
```
Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KVault.kt:229]:
```suggestion
```

Comment (gaebel) [inline src/iosMain/kotlin/com/liftric/kvault/KVault.kt:229]:
```suggestion
            kSecAttrAccessible to accessibility.value
```

**Comment by gaebel** [comment]
> Hi @gaebel thank you for the advice, I've moved the enum in KVault class and removed the optional on the property.
> For the rest, after a lot of tests I've found that the "Accessible" attributes in context method doesn't allow to save and read the property in proper way and for this I've found the solution to add it in the query of Add, Read and Update.
> I also updated the DebugViewController in order to print stored value;
> Let me know what do you think about these changes.

After having a second thought about this, I think it should only be needed to put this into the `add` and `update` methods. The read methods would only narrow the search down.  Though, I'm not sure if I want to put this into the `update` query or create another method for that matter => `set(key: String, accessibility: Accessibility)` (wouldn't be available in common). @benjohnde what do you think?

**Comment by benjohnde** [comment]
> > Hi @gaebel thank you for the advice, I've moved the enum in KVault class and removed the optional on the property.
> > For the rest, after a lot of tests I've found that the "Accessible" attributes in context method doesn't allow to save and read the property in proper way and for this I've found the solution to add it in the query of Add, Read and Update.
> > I also updated the DebugViewController in order to print stored value;
> > Let me know what do you think about these changes.
> 
> After having a second thought about this, I think it should only be needed to put this into the `add` method. The read methods would only narrow the search down. I'm not sure if I want to put this into the `update` query or create another method for that matter => `set(key: String, accessibility: Accessibility)` (wouldn't be available in common). @benjohnde what do you think?

Puh, tough question! I would solely rely on the keychain instance for handling accessibility and probably leave it in the update section as well. I would probably omit it in the first query. By doing so, you could update that accessibility value according to the current keychain instance, if and only if it is resolvable for the current device state.

**Comment by gaebel** [comment]
> @gaebel The cleaner solution would be to probably have an own set function with accessibility as parameter, but as it wouldn't be available in the common part, the current solution seems to be alright. We could stretch it and add the function in addition to the common part. So it's more the choice of the developer how to handle things on the iOS side, what do you think?

Yeah, I'd probably remove it from the `update` function and add a `set` function in the iOS target for migration purposes.

**Comment by benjohnde** [comment]
> > @gaebel The cleaner solution would be to probably have an own set function with accessibility as parameter, but as it wouldn't be available in the common part, the current solution seems to be alright. We could stretch it and add the function in addition to the common part. So it's more the choice of the developer how to handle things on the iOS side, what do you think?
> 
> Yeah, I'd probably remove it from the `update` function and add a `set` function in the iOS target for migration purposes.

Sounds great!

**Comment by cladvd** [comment]
I've removed it from update method.
About the migration, I think that should be in charge of the user following a deleting and a new add of the values that want to update with new accessibility value because also with this integration the initialized value (WhenUnlocked) is the default option already used by the system till now and will be used in the future if the accessibility property will not be changed by the user.
Is required only to test the update on the library on a device where was a previous version to check that all already saved property are accessible.
Sorry of I misunderstood some meaning 🙌

**Comment by gaebel** [comment]
> I've removed it from update method.
> About the migration, I think that should be in charge of the user following a deleting and a new add of the values that want to update with new accessibility value because also with this integration the initialized value (WhenUnlocked) is the default option already used by the system till now and will be used in the future if the accessibility property will not be changed by the user.
> Is required only to test the update on the library on a device where was a previous version to check that all already saved property are accessible.
> Sorry of I misunderstood some meaning 🙌

Yeah, this would also be a possibility. I may still add get/set functions for the accessibility in the iOS target, I can use those in a unit test to verify the changes. Just remove the accessibility from the query in the `value` method, then I'll merge your PR. Thanks again for your contribution 😊

**Comment by cladvd** [comment]
removed 😃

**Pull request merged**